From 75d9ea796f6776d00618cef8ad1fc75f4766c9c0 Mon Sep 17 00:00:00 2001
From: Hieu Huynh <hieuh@os.amperecomputing.com>
Date: Mon, 16 Sep 2024 11:29:53 +0700
Subject: [PATCH] update service: get ApplyTime from Dbus

This supports get ApplyTime from D-bus instead of hardcode to Immediate.

Signed-off-by: Hieu Huynh <hieuh@os.amperecomputing.com>
---
 redfish-core/lib/update_service.hpp | 33 ++++++++++++++++++++++++++---
 1 file changed, 30 insertions(+), 3 deletions(-)

diff --git a/redfish-core/lib/update_service.hpp b/redfish-core/lib/update_service.hpp
index 7c2c5f4c..e7e54d86 100644
--- a/redfish-core/lib/update_service.hpp
+++ b/redfish-core/lib/update_service.hpp
@@ -1129,9 +1129,36 @@ inline void
     updateSvcSimpleUpdate["TransferProtocol@Redfish.AllowableValues"] =
         std::move(allowed);
 
-    asyncResp->res
-        .jsonValue["HttpPushUriOptions"]["HttpPushUriApplyTime"]["ApplyTime"] =
-        update_service::ApplyTime::Immediate;
+    // Get the current ApplyTime value
+    sdbusplus::asio::getProperty<std::string>(
+        *crow::connections::systemBus, "xyz.openbmc_project.Settings",
+        "/xyz/openbmc_project/software/apply_time",
+        "xyz.openbmc_project.Software.ApplyTime", "RequestedApplyTime",
+        [asyncResp](const boost::system::error_code& ec,
+                    const std::string& applyTime) {
+        if (ec)
+        {
+            BMCWEB_LOG_DEBUG("DBUS response error {}", ec);
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        // Store the ApplyTime Value
+        if (applyTime == "xyz.openbmc_project.Software.ApplyTime."
+                         "RequestedApplyTimes.Immediate")
+        {
+            asyncResp->res.jsonValue["HttpPushUriOptions"]
+                                    ["HttpPushUriApplyTime"]["ApplyTime"] =
+                "Immediate";
+        }
+        else if (applyTime == "xyz.openbmc_project.Software.ApplyTime."
+                              "RequestedApplyTimes.OnReset")
+        {
+            asyncResp->res.jsonValue["HttpPushUriOptions"]
+                                    ["HttpPushUriApplyTime"]["ApplyTime"] =
+                "OnReset";
+        }
+    });
 }
 
 inline void handleUpdateServiceFirmwareInventoryCollectionGet(
-- 
2.25.1

